Development/nmos/ocsp_behaviour.cpp

```diff
@@ -0,0 +1,365 @@
+                        state.next_request = details::certificate_expiry_from_now(server_certificate_chains, gate) * 0.5;
+
+                        // construct an OCSP request with the server certificates
+                        state.ocsp_request = details::make_ocsp_request(server_certificate_chains, gate);
+
+                        slog::log<slog::severities::more_info>(gate, SLOG_FLF) << "Using the OCSP server(s):" << slog::log_manip([&](slog::log_statement& s)
+                        {
+                            for (auto& ocsp_uri : ocsp_uris)
+                            {
+                                s << '\n' << ocsp_uri.to_string();
+                            }
+                        });
+                    }
+                    catch (const nmos::experimental::ocsp_exception& e)
+                    {
+                        slog::log<slog::severities::error>(gate, SLOG_FLF) << "OCSP error during initial OCSP operation: " << e.what();
+                    }
+                    catch (const std::exception& e)
+                    {
+                        slog::log<slog::severities::error>(gate, SLOG_FLF) << "Unexpected exception during initial OCSP operation: " << e.what();
+                    }
+                    catch (...)
+                    {
+                        slog::log<slog::severities::severe>(gate, SLOG_FLF) << "Unexpected unknown exception during initial OCSP operation";
+                    }
+
+                    // implement an exponential backoff algorithm to avoid overloading the OCSP server in the event of a system restart
+                    auto lock = model.read_lock();
+                    backoff = (std::min)((std::max)((double)nmos::fields::discovery_backoff_min(model.settings), backoff * nmos::fields::discovery_backoff_factor(model.settings)), (double)nmos::fields::discovery_backoff_max(model.settings));
+                }
+                break;
+
+            case ocsp_behaviour:
+                details::ocsp_behaviour(model, ocsp_state, ocsp_uris, state, gate);
+
+                mode = initial_ocsp_operation;
+                break;
+            }
+        }
+    }
+
+    namespace details
+    {
+        web::http::client::http_client_config make_ocsp_client_config(const nmos::settings& settings, load_ca_certificates_handler load_ca_certificates, slog::base_gate& gate)
+        {
+            auto config = nmos::make_http_client_config(settings, std::move(load_ca_certificates), gate);
+            config.set_timeout(std::chrono::seconds(nmos::experimental::fields::ocsp_request_max(settings)));
+            return config;
+        }
+
+        struct ocsp_service_exception {};
+
+        // make an asynchronous POST request on the OCSP server to get certificate status
+        // see https://specs.amwa.tv/bcp-003-03/releases/v1.0.0/docs/1.0._Certificate_Provisioning.html#certificate-request
```

Comment:
I don't think this is the right reference?
If it needs a reference at all, maybe this one?
```suggestion
        // see https://specs.amwa.tv/bcp-003-01/releases/v1.0.1/docs/Secure_Communication.html#certificate-management-client
```